ipq806x: Add support for OpenMesh A42

ecsv opened the pull request:
OpenMesh A42 is a dual-band () router, based on QCA IPQ401x. More information can be found at https://wikidevi.com/wiki/Open_Mesh_A42

* QCA IPQ401x
* 256 MB of RAM
* 32 MB of SPI NOR flash (s25fl256s1)
  - 2x 15 MB available; but one of the 15 MB regions is the recovery image
* 2T2R 2.4 GHz
  - QCA4019 hw1.0 (SoC)
  - requires special BDF in QCA4019/hw1.0/board-2.bin with bus=ahb,bmi-chip-id=0,bmi-board-id=16,variant=OM-A42
* 2T2R 5 GHz
  - QCA4019 hw1.0 (SoC)
  - requires special BDF in QCA4019/hw1.0/board-2.bin with bus=ahb,bmi-chip-id=0,bmi-board-id=17,variant=OM-A42
* multi-color LED (controlled via red/green/yellow GPIOs)
* 1x button (reset; kmod-input-gpio-keys compatible)
* external watchdog
  - triggered GPIO
  - see om-watchdog
* 1x USB
* TTL pins are on board (arrow points to VCC, then follows: GND, TX, RX)
* 2x gigabit ethernet
* powered only via POE
  - 802.3af POE on Ethernet 1
  - 18-24v passive POE (mode B) on Ethernet 2

The sysupgrade image can also be used to install LEDE. The tool ap51-flash (https://dev.cloudtrax.com/git/ap51-flash.git) should be used to transfer the image to the u-boot when the device boots up.

---

The u-boot is currently only available from https://drive.google.com/drive/folders/1lagQhi80R3UMnY_8qo_5zC02gbLpSpYZ?usp=sharing

ecsv commented:
The upstreaming progress of the ath10k "qcom,ath10k-calibration-variant" patch can be found in https://patchwork.kernel.org/patch/9615185/

The upstreaming progress of the BDFs can be found in http://lists.infradead.org/pipermail/ath10k/2017-November/010470.html

chunkeey commented on target/linux/ipq806x/files-4.9/arch/arm/boot/dts/qcom-ipq4019-a42.dts:
```diff
@@ -0,0 +1,127 @@
+
+	soc {
+		wifi@a000000 {
+			status = "okay";
```
qcom-ipq4019-ap.dk01.1.dtsi enables it by default.

chunkeey commented on target/linux/ipq806x/files-4.9/arch/arm/boot/dts/qcom-ipq4019-a42.dts:
```diff
@@ -0,0 +1,127 @@
+		};
+
+		wifi@a800000 {
+			status = "okay";
```
qcom-ipq4019-ap.dk01.1.dtsi enables it by default

ecsv commented:
Thanks for the review, will change that tomorrow. I will also change the board_name (as in "$(board_name)") & BOARD_NAME) to "openmesh,a42". I will have to change how the combined-ext-image knows the CE ID but this should be fine.

mkresin commented on target/linux/ipq806x/files-4.9/arch/arm/boot/dts/qcom-ipq4019-a42.dts:
```diff
@@ -0,0 +1,127 @@
+	m25p80@0 {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		compatible = "s25fl256s1";
```
Please use only `jedec,spi-nor`. It is the standard since kernel 4.4. Everything else exists only for backwards compatibility.

ecsv commented:
I have now changed following:

* 0002-build-add-image-command-for-CE-images.patch
  - disabled use of BOARD_NAME as CE_ID/CE_TYPE
  - introduced optional parameter ce_type to overwrite the CE_TYPE (in case it should be different than the DEVICE_NAME)
* 0003-ipq806x-add-DTS-for-the-OpenMesh-A42-board.patch
  - removed `status = "okay";` from wifi nodes
  - removed `mx25l25635e@0` node
  - changed m25p80 node compatible to "jedec,spi-nor"
* 0004-ipq806x-add-user-space-support-for-OpenMesh-A42.patch
  - changed checks which compare against "$(board_name)" to "openmesh,a42" (but kept "a42:$color:status" as LED names)
  - changed $(board_name)" result to "openmesh,a42"
* 0005-ipq806x-add-sysupgrade-for-OpenMesh-A42.patch
  - changed checks which compare against "$(board_name)" to "openmesh,a42" 
* 0007-om-watchdog-add-OpenMesh-A42-support.patch
  - changed checks which compare against "$(board_name)" to "openmesh,a42" 
* 0008-ipq806x-extract-ath10k-pre-cal-for-OpenMesh-A42.patch
  - changed checks which compare against "$(board_name)" to "openmesh,a42" 
* 0009-uboot-envtools-add-OpenMesh-A42-support.patch
  - changed checks which compare against "$(board_name)" to "openmesh,a42" 
* 0012-ipq806x-create-profile-for-the-OpenMesh-A42.patch
  - changed BOARD_NAME to "openmesh,a42" 

I've also kept the package name ipq-wifi-a42. But I can also change it to ipq-wifi-openmesh-a42 when this is preferred.

ecsv commented:
The "qcom,ath10k-calibration-variant" patches were just added to ath.git https://patchwork.kernel.org/patch/10102249/

blogic commented:
With the remerge in progress, all PRs on the lede-project organisation will be closed. Please help getting this merged or rebase/post it on the openwrt project page (https://github.com/openwrt/openwrt/pulls). All remaining PRs will be closed in 30 days.

ecsv commented:
The PR can now be found in https://github.com/openwrt/openwrt/pull/633
